From fef6b9b010d3cc3752ddf6c07536ceed1fe15bee Mon Sep 17 00:00:00 2001
From: Nancy Zou <Nancy.Zou@csr.com>
Date: Wed, 27 May 2015 19:28:34 +0800
Subject: [PATCH] gl-renderer: update dma image texture every damage.

If we don't update the texture every damage, some frame will bind last frame texture.
That's for fix bug APBU-10462.

Signed-off-by: Nancy Zou <Nancy.Zou@csr.com>
---
 src/gl-renderer.c | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/src/gl-renderer.c b/src/gl-renderer.c
index 36f4385..59dc608 100644
--- a/src/gl-renderer.c
+++ b/src/gl-renderer.c
@@ -1027,9 +1027,6 @@ gl_renderer_flush_damage(struct weston_surface *surface)
 	if (!buffer)
 		return;
 
-	if (gs->buffer_type != BUFFER_TYPE_SHM) {
-		return;
-	}
 
 	/* Avoid upload, if the texture won't be used this time.
 	 * We still accumulate the damage in texture_damage, and
@@ -1049,6 +1046,14 @@ gl_renderer_flush_damage(struct weston_surface *surface)
 	if (!pixman_region32_not_empty(&gs->texture_damage) &&
 	    !gs->needs_full_upload)
 		goto done;
+	if (gs->buffer_type != BUFFER_TYPE_SHM) {
+		glActiveTexture(GL_TEXTURE0);
+		glBindTexture(gs->target, gs->textures[0]);
+		gr->image_target_texture_2d(gs->target, (EGLImageKHR)buffer->image.image);
+		buffer->image.tex = gs->textures[0];
+		return;
+	}
+
 
 	glBindTexture(GL_TEXTURE_2D, gs->textures[0]);
 
@@ -1256,10 +1261,6 @@ gl_renderer_attach_shm(struct weston_surface *es, struct weston_buffer *buffer,
 			buffer->image.pfn = free_image;
 		}
 
-		glActiveTexture(GL_TEXTURE0);
-		glBindTexture(gs->target, gs->textures[0]);
-		gr->image_target_texture_2d(gs->target, (EGLImageKHR)buffer->image.image);
-		buffer->image.tex = gs->textures[0];
 		gs->buffer_type = BUFFER_TYPE_DMA;
 	}
 }
-- 
1.9.1

